bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs

```diff
@@ -0,0 +1,318 @@
+			ensure!(asset_id == expected_asset_id, InvalidAsset);
+
+			commands.push(Command::MintForeignToken { token_id, recipient, amount });
+		}
+		Ok(commands)
+	}
+
+	/// Convert the XCM into an outbound message which can be dispatched to
+	/// the Gateway contract on Ethereum
+	///
+	/// Assets being transferred can either be Polkadot-native assets (PNA)
+	/// or Ethereum-native assets (ENA).
+	///
+	/// The XCM is evaluated in Ethereum context.
+	///
+	/// Expected Input Syntax:
+	/// ```ignore
+	/// WithdrawAsset(ETH)
+	/// PayFees(ETH)
+	/// ReserveAssetDeposited(PNA) | WithdrawAsset(ENA)
+	/// AliasOrigin(Origin)
+	/// DepositAsset(Asset)
+	/// Transact() [OPTIONAL]
+	/// SetTopic(Topic)
+	/// ```
+	/// Notes:
+	/// a. Fee asset will be checked and currently only Ether is allowed
+	/// b. For a specific transfer, either `ReserveAssetDeposited` or `WithdrawAsset` should be
+	/// 	present
+	/// c. `ReserveAssetDeposited` and `WithdrawAsset` can also be present in any order within the
+	/// 	same message
+	/// d. Currently, teleport asset is not allowed, transfer types other than
+	/// 	above will cause the conversion to fail
+	/// e. Currently, `AliasOrigin` is always required, can distinguish the V2 process from V1.
+	/// 	it's required also for dispatching transact from that specific origin.
+	/// f. SetTopic is required for tracing the message all the way along.
+	pub fn convert(&mut self) -> Result<Message, XcmConverterError> {
+		// Get fee amount
+		let fee_amount = self.extract_remote_fee()?;
+
+		// Get ENA reserve asset from WithdrawAsset.
+		let mut enas =
+			match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+		if enas.is_some() {
+			let _ = self.next();
+		}
+
+		// Get PNA reserve asset from ReserveAssetDeposited
+		let pnas = match_expression!(
+			self.peek(),
+			Ok(ReserveAssetDeposited(reserve_assets)),
+			reserve_assets
+		);
+		if pnas.is_some() {
+			let _ = self.next();
+		}
+
+		// Try to get ENA again if it is after PNA
+		if enas.is_none() {
+			enas =
+				match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+			if enas.is_some() {
+				let _ = self.next();
+			}
+		}
+		// Check AliasOrigin.
+		let origin_location = match_expression!(self.next()?, AliasOrigin(origin), origin)
+			.ok_or(AliasOriginExpected)?;
+		let origin = AgentIdOf::convert_location(origin_location).ok_or(InvalidOrigin)?;
+
+		let (deposit_assets, beneficiary) = match_expression!(
+			self.next()?,
+			DepositAsset { assets, beneficiary },
+			(assets, beneficiary)
+		)
+		.ok_or(DepositAssetExpected)?;
+
+		// assert that the beneficiary is AccountKey20.
+		let recipient = match_expression!(
+			beneficiary.unpack(),
+			(0, [AccountKey20 { network, key }])
+				if self.network_matches(network),
+			H160(*key)
+		)
+		.ok_or(BeneficiaryResolutionFailed)?;
+
+		// Make sure there are reserved assets.
+		if enas.is_none() && pnas.is_none() {
+			return Err(NoReserveAssets);
+		}
```

Comment:
Does this mean that a message that wants to do just `Transact` is not supported (pure cross-bridge contract call)?

e.g.
```
WithdrawAsset(ETH)
PayFees(ETH)
AliasOrigin(Origin)
DepositAsset(Asset)
Transact()
SetTopic(Topic)
```